### PR TITLE
Added time measurement output

### PR DESCRIPTION
**Pull request by chleh**
Sorry @wenqing that we did not respond to your time measurement PR #863. But I think time measurement is important, so here is my attempt on it.

This PR provides wall-time measurement of assembly, linear solvers, nonlinear iterations, dirchlet BCs timesteps, output, and overall runtime.

The time is measured for individual "events" and not accumulated over timesteps.

### CODE REVIEW

File: Applications/CLI/ogs.cpp [hunk 1]
```diff
@@ -74,6 +79,21 @@ int main(int argc, char *argv[])
...
     logog_setup.setLevel(log_level_arg.getValue());
 
+    BaseLib::RunTime run_time;
+    run_time.start();
+
+    {
+        auto const start_time_sys = std::chrono::system_clock::to_time_t(
+            std::chrono::system_clock::now());
+        std::ostringstream sstr;
+        sstr << std::put_time(std::localtime(&start_time_sys), "%F %T %z");
```
Comment (wenqing) [inline Applications/CLI/ogs.cpp]: % before T and Z display in red colour, but the expression looks nothing wrong.

**Comment by wenqing** [comment]
:+1:

File: Applications/CLI/ogs.cpp [hunk 2]
```diff
@@ -74,6 +77,23 @@ int main(int argc, char *argv[])
...
+        auto const start_time_sys = std::chrono::system_clock::to_time_t(
+            std::chrono::system_clock::now());
+        char time_str[100];
+        if (std::strftime(time_str, sizeof(time_str), "%F %T %z",
+            std::localtime(&start_time_sys))) {
+	    INFO("OGS started on %s.", time_str);
+	}
+    }
+
+    std::chrono::steady_clock::now();
```
Comment (TomFischer) [inline Applications/CLI/ogs.cpp]: Since the return value is not stored here I wonder if this line is necessary?

File: Applications/CLI/ogs.cpp [hunk 3]
```diff
@@ -74,6 +77,23 @@ int main(int argc, char *argv[])
...
+
+    {
+        auto const start_time_sys = std::chrono::system_clock::to_time_t(
+            std::chrono::system_clock::now());
+        char time_str[100];
+        if (std::strftime(time_str, sizeof(time_str), "%F %T %z",
+            std::localtime(&start_time_sys))) {
+	    INFO("OGS started on %s.", time_str);
+	}
+    }
```
Comment (TomFischer) [inline Applications/CLI/ogs.cpp]: Maybe lines 84-90 can be moved as a general function into BaseLib?

File: Applications/CLI/ogs.cpp [hunk 4]
```diff
@@ -117,9 +137,22 @@ int main(int argc, char *argv[])
...
     }
+
+    {
+        auto const end_time_sys = std::chrono::system_clock::to_time_t(
+            std::chrono::system_clock::now());
+        char time_str[100];
+        if (std::strftime(time_str, sizeof(time_str), "%F %T %z",
+            std::localtime(&end_time_sys))) {
+	    INFO("OGS terminated on %s.", time_str);
+	}
```
Comment (TomFischer) [inline Applications/CLI/ogs.cpp]: Here you could use the suggested general BaseLib function and avoid code duplication.

**Comment by TomFischer** [comment]
I do not know if it would be a good idea to implement a new loging macro TIME into logog for timing output. What do you think?

**Comment by TomFischer** [comment]
After moving extracting code to BaseLib: :+1:

**Comment by chleh** [comment]
@wenqing: I'd prefer not to accumulate in ogs. I think such accumulation can be easily done by grepping the measured times from the log and summing them up in a script.

**Comment by chleh** [comment]
@TomFischer: Since the logog macros and logog log line prefixes are defined in macro.hpp and formatter.cpp in that third party package I don't want to introduce a custom TIME log macro now.

**Comment by chleh** [comment]
In the last commit I added functionality that prints the OGS version in every run.

File: BaseLib/DateTools.h [hunk 5]
```diff
@@ -55,6 +57,15 @@ int strDate2int(const std::string &s);
  */
 int xmlDate2int(const std::string &s);
 
+/**
+ * Formats the given time point according to RFC 3339 (cf. man-page of the unix
+ * date utility).
+ *
+ * Example: 2006-08-14 02:34:56-06:00
```
Comment (endJunction) [inline BaseLib/DateTools.h]: I'd prefer not to print the year and the timezone, which are very unlikely to change in a single run...

Comment (chleh) [inline BaseLib/DateTools.h]: I'd rather keep both of them. `08-14 02:34:56` doesn't look like a real timestamp, does it? :smile:

**Pull request merged**